beacon-chain/db/block.go

```diff
@@ -35,6 +47,15 @@ func createBlock(enc []byte) (*pb.BeaconBlock, error) {
 // Block accepts a block root and returns the corresponding block.
 // Returns nil if the block does not exist.
 func (db *BeaconDB) Block(root [32]byte) (*pb.BeaconBlock, error) {
+	db.blocksLock.Lock()
+	defer db.blocksLock.Unlock()
+
+	// Return block from cache if it exists
+	if _, exists := db.blocks[root]; exists {
+		blockCacheHit.Inc()
+		return db.blocks[root], nil
```

Comment:
I think you should take a read lock until you have a miss.

```go
db.blocksLock.RLock()

if exists ... {
   db.blocksLock.RUnlock()
   return block
}
db.blocksLock.RUnlock()
db.blocksLock.Lock()
defer db.blocksLock.Unlock()
```

If you have a really high hit ratio, there is no reason you couldn't have multiple readers accessing the value at the same time. What do you think?